coins/ethereum/contracts/Router.sol

```diff
@@ -0,0 +1,101 @@
+//SPDX-License-Identifier: AGPLv3
+pragma solidity ^0.8.0;
+
+import "./Schnorr.sol";
+
+contract Router is Schnorr {
+    // contract owner
+    address owner;
+
+    // nonce is incremented for each batch of transactions executed
+    uint256 nonce; 
+
+    // prevents re-entrancy
+    uint8 internal locked;
+
+    struct PublicKey {
+        uint8 parity;
+        bytes32 px;
+    }
+
+    // current aggregated validator public key 
+    PublicKey publicKey;
+
+    struct Transaction {
+        address to;
+        uint256 value;
+        uint256 gas;
+        bytes data;
+    }
+
+    struct Signature {
+        bytes32 e;
+        bytes32 s;
+    }
+
+    event Executed(uint256 nonce, uint256 index, bool success);
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "must be called by the contract owner");
+        _;
+    }
+
+    modifier noReentrant() {
+        require(locked == 0, "no re-entrancy");
+        locked = 1;
+        _;
+        locked = 0;
+    }
+
+    function getNonce() external view returns (uint256) {
+        return nonce;
+    }
+
+    // setPublicKey can be called by the contract owner to set the current public key,
+    // only if the public key has not been set.
+    function setPublicKey(
+        PublicKey memory _publicKey
+    ) public onlyOwner {
+        require(publicKey.px == 0, "public key has already been set");
+        publicKey.parity = _publicKey.parity;
```

Comment:
Can we not assign the struct as a whole?

Reply:
it doesn't seem to get set if I just set the struct (maybe bc it's declared as memory)? seems weird though, I can retest